cs/src/comm/epoxy-transport/EpoxyNetworkStream.cs

```diff
@@ -58,6 +58,62 @@ private EpoxyNetworkStream(Socket socket, Stream stream, Logger logger)
             this.logger = logger;
         }
 
+        /// <summary>
+        /// Safely creates an EpoxyNetworkStream from a TCP socket, handling cleaning up resources
+        /// when errors are encountered.
+        /// </summary>
+        /// <remarks>
+        /// This is a helper function that centralizes error handling and socket shutdown.
+        /// </remarks>
+        /// <param name="getSocketFunc">A function to invoke to acquire a socket.</param>
+        /// <param name="getNetworkStreamFunc">
+        /// A function to invoke to wrap a socket in a network stream.
+        /// </param>
+        /// <param name="timeoutConfig">The timeout config to use for this stream.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns>A connected EpoxyNetworkStream.</returns>
+        public static async Task<EpoxyNetworkStream> SocketToNetworkStreamAsync(
+            Func<Task<Socket>> getSocketFunc,
+            Func<Socket, Task<EpoxyNetworkStream>> getNetworkStreamFunc,
+            EpoxyTransport.TimeoutConfig timeoutConfig,
+            Logger logger)
+        {
+            Socket socket = null;
+            EpoxyNetworkStream epoxyStream = null;
+
+            try
+            {
+                socket = await getSocketFunc();
+                ConfigureSocketKeepAlive(socket, timeoutConfig, logger);
```

Comment:
I just confirmed that .NET (on Windows at least) can reconfigure the socket keep alive times after the socket has been configured.